### DISCUSSION TITLE
Stack overflow when trying to remove an item from a treeview

### DISCUSSION THREAD

**Discussion by ffquintella** [AvaloniaUI/Avalonia#12692] [Q&A]
I get a stack overflow error every time I try to remove and Item from a collection binded to my tree view. I debugged and the problem seems to be inside one of the avalonia classes. 

This is the tree view 

            
            <TreeView Grid.Row="1" ItemsSource="{Binding Nodes, Mode=TwoWay}"
                      SelectedItem="{Binding SelectedNode, Mode=TwoWay}">
                <TreeView.ItemTemplate>
                    <TreeDataTemplate ItemsSource="{Binding SubNodes, Mode=TwoWay}" >
                        <TextBlock Text="{Binding Title}" IsVisible="{Binding IsVisible, Mode=TwoWay}"/>
                    </TreeDataTemplate>
                </TreeView.ItemTemplate>
            </TreeView>

This is the treenode class:
public class TreeNode: ReactiveObject
{
    
    private ObservableCollection<TreeNode>? _subNodes;

    public ObservableCollection<TreeNode>? SubNodes
    {
        get => _subNodes;
        set
        {
            if (value == null) return;
            this.RaiseAndSetIfChanged(ref _subNodes, value);
        }
    }
    
    public string Title { get; }
    
    public int EntityId { get; }

    public bool IsVisible { get; set; } = true;

    public TreeNode(string title)
    {
        Title = title;
    }

    public TreeNode(string title, int entityId, ObservableCollection<TreeNode> subNodes)
    {
        EntityId = entityId;
        Title = title;
        SubNodes = subNodes;
    }
}

And this is the instruction that starts the problem:

Nodes!.Remove(rootNode);

**To Reproduce**
I guess it would happen with a treeview with similar structure while trying to remove any item.

I am using avalonia 11.0.4

**Comment by IanRawley** [comment]
Is there a reason that your ItemsSource and the TextBlock's IsVisible bindings are TwoWay?  I'm not sure how that would cause a StackOverflow exception rather than an NRE, but it does look strange to me.

**Comment by timunie** [comment]
Yeah, use two-way binding only if you know you need to write back things that the user set in the view to the ViewModel. Example:
```
TextBox.Text
CheckBox.IsChecked
ToggleButton.IsChecked
```

but the user cannot change `IsVisible`, `ItemsSource`, ... -> so these properties are one-way bindings. Only the ViewModel can change it and inform the View.

Aynway, I tried  your provided sample and I'm unable to reproduce. So please upload a very minimal App to reproduce. Thanks 🙏

**Comment by timunie** [comment]
@ffquintella I'm moving this into Q&A because I think there is something going on in your ViewModel (as I can't reproduce). If we get your minimal sample and it is an Avalonia bug, we will open a new issue with detailed steps to reproduce, okay?

**Comment by ffquintella** [comment]
I only used two-way binding as a form of testing, the problem happens on the one-way binding also.

I think the best way to reproduce is checkout the project here https://github.com/ffquintella/netrisk It´s happening on the GUIClient and happens only when I try to remove one item from the source collection.

**Reply by timunie** [reply]
sorry, but we don't have the capacity to check such huge solutions (the entire source must be scanned for harmful code). We need minimal samples if possible. Thx for your understanding.

**Reply by ffquintella** [reply]
No need I found a way to make it behaves like I need.

**Comment by ffquintella** [comment]
I finally found a workaround to solve my problem. But debugging, basically there is a stack overflow when using the two-way binding. This should either be threaded to a specific exception or forbidden.

Any way, thanks for the feedback guys.

**Reply by stevemonaco** [reply]
I have similar tree node removal operations in my app and they work fine. I would suspect some re-entrant cycle between your removing of a node and removing of an entity. You need to be quite careful of this in systems where changes are pushed/pulled. You could set a breakpoint in both methods and see if they continually trigger each other and perhaps watch the call stack grow.
